warehouse/oidc/models.py

```diff
@@ -0,0 +1,168 @@
+    # not checked as part of verifying the JWT.
+    __unchecked_claims__ = set()
+
+    def verify_claims(self, signed_claims):
+        """
+        Given a JWT that has been successfully decoded (checked for a valid
+        signature and basic claims), verify it against the more specific
+        claims of this provider.
+        """
+
+        # Defensive programming: treat the absence of any claims to verify
+        # as a failure rather than trivially valid.
+        if not self.__verifiable_claims__:
+            return False
+
+        # All claims should be accounted for.
+        # The presence of an unaccounted claim is not an error, only a warning
+        # that the JWT payload has changed.
+        known_claims = self.__verifiable_claims__.keys().union(
+            self.__preverified_claims__, self.__unchecked_claims__
+        )
+        unaccounted_claims = known_claims.difference(signed_claims.keys())
+        if unaccounted_claims:
+            sentry_sdk.capture_message(
+                f"JWT for {self.__class__.__name__} has unaccounted claims: "
+                f"{unaccounted_claims}"
+            )
+
+        # Finally, perform the actual claim verification.
+        for claim_name, check in self.__verifiable_claims__.items():
+            if not check(self.getattr(claim_name), signed_claims[claim_name]):
+                return False
+
+        return True
+
+    @property
+    def provider_name(self):
+        # Only concrete subclasses of OIDCProvider are constructed.
+        return NotImplemented
+
+
+class GitHubProvider(OIDCProvider):
+    __tablename__ = "github_oidc_providers"
+    __mapper_args__ = {"polymorphic_identity": "github_oidc_providers"}
+    __table_args__ = (
+        UniqueConstraint(
+            "repository_name",
+            "owner",
+            "workflow_name",
+            name="_github_oidc_provider_uc",
+        ),
+    )
+
+    id = Column(UUID(as_uuid=True), ForeignKey(OIDCProvider.id), primary_key=True)
+    repository_name = Column(String)
+    owner = Column(String)
+    owner_id = Column(String)
+    workflow_name = Column(String)
+
+    __verifiable_claims__ = {
+        "repository": str.__eq__,
+        "job_workflow_ref": str.startswith,
+        "actor": str.__eq__,
+        "workflow": str.__eq__,
+    }
+
+    __unchecked_claims__ = {
+        "jti",
+        "sub",
+        "ref",
+        "sha",
+        "run_id",
+        "run_number",
+        "run_attempt",
+        "head_ref",
+        "base_ref",
+        "event_name",
+        "ref_type",
+    }
+
+    @property
+    def provider_name(self):
+        return "GitHub"
+
+    @property
+    def repository(self):
+        return f"{self.owner}/{self.repository_name}"
+
+    @property
+    def job_workflow_ref(self):
+        return f"{self.repository}/.github/workflows/{self.workflow_name}.yml"
+
+    @property
+    def actor(self):
+        return self.owner
```

Comment:
Is this right? Will the actor always be the repo owner?

Reply:
Hmm, nope -- the `actor` can be anybody empowered to trigger the workflow. This should probably become one of our unvalidated claims, unless we want to enforce an allowlist of users here.

Reply:
2c9722d4